### ISSUE TITLE
Refactor UrlPath

### ISSUE THREAD

**Issue by madwort** [opensafely-core/airlock#622]
> There are times when a UrlPath means "path to file in a workspace" and times when it means "actual url" and times where it means "request url that will include group". cf bll.get_request_file_from_output_path and bll.get_request_file_from_urlpath.
> 
> Previously, I tried to split those into separate types: OutputPath, UrlPath, RequestUrlPath (which had a .group as you suggest and a .output_path w/o the group), and use the right types to help us catch these distintions.
> 
> But at the time, I couldn't make it work cleanly. It's probably worth trying again, and definitely would be a good refactor.
> 

https://github.com/opensafely-core/airlock/pull/621#discussion_r1718240397

**Comment by madwort** [comment]
I think I've fought this a bit in the past (having to keep in mind whether a path contains a group name or not) - would be great to have it really obvious (& enforced by the type system).

**Comment by bloodearnest** [comment]
One idea I wanted to try here was have the types enforced/converted in our django urls.py patterns, so that the views got typed instances of UrlPath/GroupPath, rather than strings.

You can do this with custom url type converters easily enough: https://docs.djangoproject.com/en/5.0/topics/http/urls/#registering-custom-path-converters

**Comment by madwort** [comment]
I'm not sure we can do it with custom path converters if we need access to other data, e.g. `request.user`, `release_request` etc - we could do it with middleware though?

**Comment by bloodearnest** [comment]
I don't think we need anything except path components for a converter to UrlPath/GroupPath?

**Comment by madwort** [comment]
actually you're right - I thought this was going to verify filenames/groupnames against a request but for these purposes it doesn't. this has given me the idea to do a similar thing to other params, though, which do need these things - e.g. release_request (see slack thread https://bennettoxford.slack.com/archives/C069YDR4NCA/p1741949840761249 )

**Comment by rw251** [comment]
When I was working on allowing groups to be renamed, I did [some preliminary work](https://github.com/opensafely-core/airlock/compare/main...rw/separate-group-from-path) that separated the `group_name` field from the `path` in the URLs (the next step would have been `group_name` -> `group_id`. This may or may not be useful depending on your approach, but linking it here so at least you know about it.
